# SV-LRvsSR

A study to compare structural variation (SV) predictions from 10X Genomics linked-reads sequencing (10XWGS) and conventional Illumina short-reads sequencing (cWGS). We predict SVs from Long Ranger pipeline of 10X Genomics, NAIBR and GROC-SV for linked-reads. And we predict SV from Delly, Lumpy and SvABA and combine the predictions with linked-read tools. The MCF7 breast cancer cell line and a primary breast tumor sample was analysed for the comparison. Apart from comparing different predictions by two technologies we also propose precision enrichment logistic regression models for retaining true positive SVs amongst the high false positive calls from both the technologies.

The code for analysis was written in python (version 2.7.15) and precision enrichment models were trained and built in R (version 3.6). The commands for running all tools and analysis pipelines are mentioned in Commands.org. The codes are present in folder Script and trained models for enrichment are in Model folder.
